internal/mode/static/state/graph/policies.go

```diff
@@ -0,0 +1,310 @@
+
+	for _, policy := range g.NGFPolicies {
+		ref := policy.TargetRef
+
+		switch ref.Kind {
+		case kinds.Gateway:
+			attachPolicyToGateway(policy, g.Gateway, g.IgnoredGateways, ctlrName)
+		case kinds.HTTPRoute, kinds.GRPCRoute:
+			route, exists := g.Routes[routeKeyForKind(ref.Kind, ref.Nsname)]
+			if !exists {
+				continue
+			}
+
+			attachPolicyToRoute(policy, route, ctlrName)
+		}
+	}
+}
+
+func attachPolicyToRoute(policy *Policy, route *L7Route, ctlrName string) {
+	kind := v1.Kind(kinds.HTTPRoute)
+	if route.RouteType == RouteTypeGRPC {
+		kind = kinds.GRPCRoute
+	}
+
+	routeNsName := types.NamespacedName{Namespace: route.Source.GetNamespace(), Name: route.Source.GetName()}
+
+	ancestor := &PolicyAncestor{
+		Ancestor: createParentReference(v1.GroupName, kind, routeNsName),
+	}
+
+	curAncestorStatus := policy.Source.GetPolicyStatus().Ancestors
+	if ancestorsFull(curAncestorStatus, ancestor.Ancestor, ctlrName) {
+		return
+	}
+
+	policy.Ancestor = ancestor
+
+	if !route.Valid || !route.Attachable || len(route.ParentRefs) == 0 {
+		policy.Ancestor.Conditions = []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is invalid")}
+
+		return
+	}
+
+	route.Policies = append(route.Policies, policy)
+}
+
+func attachPolicyToGateway(
+	policy *Policy,
+	gw *Gateway,
+	ignoredGateways map[types.NamespacedName]*v1.Gateway,
+	ctlrName string,
+) {
+	ref := policy.TargetRef
+
+	_, ignored := ignoredGateways[ref.Nsname]
+
+	if !ignored && ref.Nsname != client.ObjectKeyFromObject(gw.Source) {
+		return
+	}
+
+	ancestor := &PolicyAncestor{
+		Ancestor: createParentReference(v1.GroupName, kinds.Gateway, ref.Nsname),
+	}
+
+	curAncestorStatus := policy.Source.GetPolicyStatus().Ancestors
+	if ancestorsFull(curAncestorStatus, ancestor.Ancestor, ctlrName) {
+		return
+	}
+
+	policy.Ancestor = ancestor
+
+	if ignored {
+		policy.Ancestor.Conditions = []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is ignored")}
+		return
+	}
+
+	if !gw.Valid {
+		policy.Ancestor.Conditions = []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is invalid")}
+		return
+	}
+
+	gw.Policies = append(gw.Policies, policy)
+}
+
+func processPolicies(
+	policies map[PolicyKey]policies.Policy,
+	validator validation.PolicyValidator,
+	gateways processedGateways,
+	routes map[RouteKey]*L7Route,
+) map[PolicyKey]*Policy {
+	if len(policies) == 0 || gateways.Winner == nil {
+		return nil
+	}
+
+	processedPolicies := make(map[PolicyKey]*Policy)
+
+	for key, policy := range policies {
+		ref := policy.GetTargetRef()
+		refNsName := targetRefNsName(ref, policy.GetNamespace())
+
+		refGroupKind := fmt.Sprintf("%s/%s", ref.Group, ref.Kind)
+
+		switch refGroupKind {
+		case gatewayGroupKind:
+			if !gatewayExists(refNsName, gateways.Winner, gateways.Ignored) {
+				continue
+			}
+		case hrGroupKind, grpcGroupKind:
+			if _, exists := routes[routeKeyForKind(ref.Kind, refNsName)]; !exists {
+				continue
+			}
+		default:
+			continue
```

Comment:
Just to clarify, this filters out any policies that reference resources that we aren't tracking. So in that case, we aren't adding any conditions to the policy to say "RefNotFound", correct? We only add conditions if we know the policy references a resource we care about but there is some error (determined in attachPolicies).

Reply:
Yes, that's correct. Nick talks about that here: https://github.com/kubernetes-sigs/gateway-api/issues/2755#issuecomment-1970397263.